File: packages/depz-sensor-sdk-cpp/src/identity.cpp
#include "depz/identity.hpp"

#include <regex>

namespace depz {

std::string to_string(SensorType t) {
    switch (t) {
        case SensorType::Sr04: return "sr04";
        case SensorType::Vl53l8: return "vl53l8";
        case SensorType::Bno086: return "bno086";
        case SensorType::Unknown: return "unknown";
    }
    return "unknown";
}

std::string to_string(DeviceMode m) {
    switch (m) {
        case DeviceMode::App: return "app";
        case DeviceMode::Bootloader: return "bootloader";
        case DeviceMode::Unknown: return "unknown";
    }
    return "unknown";
}

Identity parse_software_name(const std::string& name) {
    static const std::regex version_re(R"(_v(\d+(?:\.\d+)*)$)");

    std::string version;
    std::smatch m;
    if (std::regex_search(name, m, version_re)) {
        version = m[1].str();
    }

    auto starts_with = [&](const char* prefix) {
        std::string p(prefix);
        return name.size() >= p.size() && name.compare(0, p.size(), p) == 0;
    };
    auto contains = [&](const char* tok) { return name.find(tok) != std::string::npos; };

    if (starts_with("BOOTDEPZ")) {
        return Identity{DeviceMode::Bootloader, std::nullopt, name, version};
    }
    if (starts_with("APP_")) {
        if (contains("SR04")) return Identity{DeviceMode::App, SensorType::Sr04, name, version};
        if (contains("VL53L8")) return Identity{DeviceMode::App, SensorType::Vl53l8, name, version};
        if (contains("BNO086")) return Identity{DeviceMode::App, SensorType::Bno086, name, version};
        return Identity{DeviceMode::App, SensorType::Unknown, name, version};
    }
    return Identity{DeviceMode::Unknown, std::nullopt, name, version};
}

}  // namespace depz

File: packages/depz-sensor-sdk-rust/src/protocol/identity.rs
//! Firmware-name parsing (contracts/02_COMMON_COMMANDS.md §4).

/// Product family classified from the active-software name.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SensorType {
    Sr04,
    Vl53l4,
    Vl53l8,
    Bno086,
    Unknown,
}

impl SensorType {
    /// Lowercase wire-string form (matches the golden vectors).
    pub fn as_str(&self) -> &'static str {
        match self {
            SensorType::Sr04 => "sr04",
            SensorType::Vl53l4 => "vl53l4",
            SensorType::Vl53l8 => "vl53l8",
            SensorType::Bno086 => "bno086",
            SensorType::Unknown => "unknown",
        }
    }
}

/// Firmware operating mode.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Mode {
    App,
    Bootloader,
    Unknown,
}

impl Mode {
    pub fn as_str(&self) -> &'static str {
        match self {
            Mode::App => "app",
            Mode::Bootloader => "bootloader",
            Mode::Unknown => "unknown",
        }
    }
}

/// Classified `GET_NAME_ACTIVE_SOFTWARE` result.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Identity {
    pub mode: Mode,
    /// `None` in bootloader/unknown mode.
    pub sensor_type: Option<SensorType>,
    pub software_name: String,
    /// `""` when not parseable.
    pub version: String,
}

const PRODUCT_TOKENS: &[(&str, SensorType)] = &[
    ("SR04", SensorType::Sr04),
    ("VL53L4", SensorType::Vl53l4),
    ("VL53L8", SensorType::Vl53l8),
    ("BNO086", SensorType::Bno086),
];

/// Match `_v(\d+(\.\d+)*)$`: find the leftmost `_v` whose entire remainder is a
/// dotted numeric version ending at the string end. Returns the captured version.
fn extract_version(name: &str) -> String {
    let bytes = name.as_bytes();
    if bytes.len() < 2 {
        return String::new();
    }
    for i in 0..=bytes.len() - 2 {
        if &bytes[i..i + 2] == b"_v" {
            let rest = &name[i + 2..];
            if is_dotted_version(rest) {
                return rest.to_string();
            }
        }
    }
    String::new()
}

/// `\d+(\.\d+)*` matched against the whole string.
fn is_dotted_version(s: &str) -> bool {
    if s.is_empty() {
        return false;
    }
    let mut prev_dot = true; // require a digit at the start of each group
    for c in s.chars() {
        if c == '.' {
            if prev_dot {
                return false; // empty group / leading dot / double dot
            }
            prev_dot = true;
        } else if c.is_ascii_digit() {
            prev_dot = false;
        } else {
            return false;
        }
    }
    !prev_dot // must not end on a dot
}

/// Classify an already-stripped active-software string.
pub fn parse_software_name(name: &str) -> Identity {
    let version = extract_version(name);
    if name.starts_with("BOOTDEPZ") {
        return Identity {
            mode: Mode::Bootloader,
            sensor_type: None,
            software_name: name.to_string(),
            version,
        };
    }
    if name.starts_with("APP_") {
        for (token, sensor) in PRODUCT_TOKENS {
            if name.contains(token) {
                return Identity {
                    mode: Mode::App,
                    sensor_type: Some(*sensor),
                    software_name: name.to_string(),
                    version,
                };
            }
        }
        return Identity {
            mode: Mode::App,
            sensor_type: Some(SensorType::Unknown),
            software_name: name.to_string(),
            version,
        };
    }
    Identity {
        mode: Mode::Unknown,
        sensor_type: None,
        software_name: name.to_string(),
        version,
    }
}
